components/webext-storage/src/api.rs

```diff
@@ -0,0 +1,425 @@
+    // value (but presumably that's going to be the uncommon case, so it probably
+    // doesn't matter)
+    let val_map = match val {
+        JsonValue::Object(m) => m,
+        // Not clear what the error semantics should be yet. For now, pretend an empty map.
+        _ => Map::new(),
+    };
+
+    let mut current = get_from_db(conn, ext_id)?.unwrap_or_else(Map::new);
+
+    let mut changes = StorageChanges::with_capacity(val_map.len());
+
+    // iterate over the value we are adding/updating.
+    for (k, v) in val_map.into_iter() {
+        let old_value = current.remove(&k);
+        if current.len() >= MAX_ITEMS {
+            return Err(ErrorKind::QuotaError(QuotaReason::MaxItems).into());
+        }
+        // Sadly we need to stringify the value here just to check the quota.
+        // Reading the chrome docs literally, the length of the key is just
+        // the string len, but the value is the json val.
+        if k.bytes().count() + v.to_string().bytes().count() >= QUOTA_BYTES_PER_ITEM {
+            return Err(ErrorKind::QuotaError(QuotaReason::ItemBytes).into());
+        }
+        current.insert(k.clone(), v.clone());
+        let change = StorageValueChange {
+            old_value,
+            new_value: Some(v),
+        };
+        changes.insert(k, change);
+    }
+
+    save_to_db(conn, ext_id, &JsonValue::Object(current))?;
+    Ok(changes)
+}
+
+// A helper which takes a param indicating what keys should be returned and
+// converts that to a vec of real strings. Also returns "default" values to
+// be used if no item exists for that key.
+fn get_keys(keys: &JsonValue) -> Vec<(String, Option<JsonValue>)> {
+    match keys {
+        JsonValue::String(s) => vec![(s.to_string(), None)],
+        JsonValue::Array(keys) => {
+            // because nothing with json is ever simple, each key may not be
+            // a string. We ignore any which aren't.
+            keys.iter()
+                .filter_map(|v| v.as_str().map(|s| (s.to_string(), None)))
+                .collect()
+        }
+        // XXX - we clone the map value here, but `remove()` doesn't need it - maybe
+        // we should take a param to indicate if the defaults are actually needed?
+        // (Or maybe lifetimes magic could make the clone unnecessary? It should have
+        // the same lifetime as `keys`)
+        JsonValue::Object(m) => m
+            .iter()
+            .map(|(k, d)| (k.to_string(), Some(d.clone())))
+            .collect(),
+        _ => vec![],
+    }
+}
+
+/// The implementation of `storage[.sync].get()` - on success this always
+/// returns a Json object.
+pub fn get(conn: &Connection, ext_id: &str, keys: &JsonValue) -> Result<JsonValue> {
+    // key is optional, or string or array of string or object keys
+    let maybe_existing = get_from_db(conn, ext_id)?;
+    let mut existing = match maybe_existing {
+        None => return Ok(JsonValue::Object(Map::new())),
```

Comment:
I'm surprised this returns an empty object and not `JsonValue::Null`!

Reply:
Me too - I discovered this when running desktop tests - but it makes sense when you think that get can actually request any number of keys